lib/dx/loaders/dataloader.ex

```diff
@@ -49,20 +49,21 @@ defmodule Dx.Loaders.Dataloader do
   defp where(opts, []), do: opts
   defp where(opts, conditions), do: Keyword.put(opts, :where, {:all, conditions})
 
-  def init() do
+  def init(opts \\ []) do
     repo = config(:repo)
 
     # workaround for dataloader incompatibility with transactions
     #   -> https://github.com/absinthe-graphql/dataloader/issues/129#issuecomment-965492108
     run_concurrently? = not db_conn_checked_out?(repo)
 
-    source =
-      Dataloader.Ecto.new(repo,
-        query: &Dx.Ecto.Query.from_options/2,
-        async: run_concurrently?
-      )
+    default_opts = [get_policy: :tuples, async: run_concurrently?]
+    opts = Keyword.merge(default_opts, opts[:dataloader] || [])
+    default_ecto_opts = [query: &Dx.Ecto.Query.from_options/2, async: run_concurrently?]
+    ecto_opts = Keyword.merge(default_ecto_opts, opts[:dataloader_ecto] || [])
```

Comment:
Just like for `Dataloader.new`, `query` is required for the `Dx` implementation and should not be overridable. `async` should not be overridable either (see above).

The only relevant options for the Ecto-specific `dataloader` source are `:repo_opts` and `:timeout` ([code](https://github.com/absinthe-graphql/dataloader/blob/fcd81aa911eadb41a126cb39bf9580c7e99755ed/lib/dataloader/ecto.ex#L282-L283)).

- `repo_opts` is forwarded to `Ecto.Repo` functions, that would be good to support. I'd call it `repo_options` though for consistency on Dx's side.
- `timeout` is the timeout per `dataloader` source. That's what we want to be able to set.

I think it would be fine to have them at the root options level though, e.g.

```elixir
Dx.load!(list, :state,
  loader_options: [
    timeout: :timer.seconds(25),
    repo_options: [telemetry_options: [dx: true]]
  ]
)
```

Or maybe even without the `loader_options`, since we only support these two options? 🤔

Reply:
Yes, fair point, I agree about keeping the `loader_options` 👍

I do like the idea of passing `loader: {Module, opts: ...}`. However, it should be convenient to use the default loader with custom options. And also (if you agree) easy to inspect the options that were passed, maybe independent of the loader implementation. Both would speak for a separate `loader_options` in the `Dx.Evaluation` struct.

In the future (or now), we could allow passing `Dx.load!(list, :state, loader: {Module, opts: ...})` and split it in `add_options`.

What do you think?

Reply:
Of course, we could do it the other way around, storing the `{Module, opts}` in the struct and splitting it on initialising the loader. I'd lean toward storing it in separate fields, because I think it makes the implementation a bit simpler. But it's up to you.

Reply:
Lets just keep it simple then for now:
Pass as separate fields. Store as separate fields.

Only downside: The types aren't quite as nice, as we lose the connection between loader type and loader_options type. But I think simplicity outweighs that.